server/src/main/java/org/opensearch/identity/authz/WriteableParameter.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    private CheckableParameter<?> parameter;
+    private ParameterType type;
+
+    /**
+     * Wrap a {@link CheckableParameter}. The generic type is determined from the type of the value.
+     *
+     * @param parameter  The parameter to wrap. The value must not be null.
+     * @throws IllegalArgumentException if the parameter has a null value.
+     */
+    public WriteableParameter(CheckableParameter<?> parameter) {
+        this(parameter, ParameterType.valueOf(parameter.getType().getSimpleName()));
+    }
+
+    /**
+     * Wrap a {@link CheckableParameter} with a specified generic type.
+     *
+     * @param parameter  The parameter to wrap.
+     * @param type  The Generic type of the parameter.
+     */
+    public WriteableParameter(CheckableParameter<?> parameter, ParameterType type) {
+        this.parameter = parameter;
+        this.type = type;
+    }
+
+    /**
+     * Wrap a {@link CheckableParameter} read from a stream.
+     *
+     * @param in Input to read the value from.
+     * @throws IOException if there is an error reading the values
+     */
+    public WriteableParameter(StreamInput in) throws IOException {
+        // Read the type
+        this.type = in.readEnum(ParameterType.class);
+        // Read the key
+        String key = in.readString();
+        // Read the value
+        Object value = readValue(in);
+        this.parameter = createParameter(type, key, value);
+    }
+
+    /**
+     * Gets the wrapped setting. Use {@link #getType()} to determine its generic type.
+     *
+     * @return The wrapped setting.
+     */
+    public CheckableParameter<?> getParameter() {
+        return this.parameter;
+    }
+
+    /**
+     * Gets the generic type of the wrapped setting.
+     *
+     * @return The wrapped setting's generic type.
+     */
+    public ParameterType getType() {
+        return this.type;
+    }
+
+    @SuppressWarnings("unchecked")
+    private CheckableParameter<?> createParameter(ParameterType type, String key, Object value) {
+        switch (type) {
+            case Boolean:
+                return new CheckableParameter<>(key, (Boolean) value, Boolean.class);
+            case Integer:
+                return new CheckableParameter<>(key, (Integer) value, Integer.class);
+            case Long:
+                return new CheckableParameter<>(key, (Long) value, Long.class);
+            case Float:
+                return new CheckableParameter<>(key, (Float) value, Float.class);
+            case Double:
+                return new CheckableParameter<>(key, (Double) value, Double.class);
+            case String:
+                return new CheckableParameter<>(key, (String) value, String.class);
+            case TimeValue:
+                return new CheckableParameter<>(key, (TimeValue) value, TimeValue.class);
+            case ByteSizeValue:
+                return new CheckableParameter<>(key, (ByteSizeValue) value, ByteSizeValue.class);
+            default:
```

Comment:
Are we planning on supporting custom user object types here? (i.e. have a `case` that supports a generic type)

Reply:
No, I expect we will define a list of parameters that are checkable. We need to support index patterns to support existing index permissions. Other parameters that are needed to perform authorization could be resource identifiers.